app/controllers/api/events_controller.rb

```diff
@@ -0,0 +1,55 @@
+module Api
+  class EventsController < ApiController
+    PERMITTED_EVENT_PARAMS = [
+      :type,
+      attributes: [
+        :event_type,
+        :client_timestamp,
+        :notes,
+        details: {},
+      ],
+      relationships: {},
+    ].freeze
+
+    def create
+      Events::CommonParamsValidator.new(event_params).validate!
+      event = Event.create(event_attributes)
+
+      render json: event, status: :created, serializer: ::EventSerializer
+    end
+
+  private
+
+    def event_attributes
+      {}.tap do |attributes|
+        attributes.merge!(event_params.fetch(:attributes, {}))
+
+        # NB: json:api doesn't like the rails convention to name the STI column as type. We require 'event_type' in the request but the column is actually type
+        attributes.delete('event_type')
+
+        attributes.merge!('type' => event_type)
+        attributes.merge!('event_name' => Event::NA)
+        attributes.merge!('eventable' => eventable) if eventable_params
+      end
+    end
```

Comment:
I think it would be more confusing at first glance to have attribute merging done in more than one location, though. It's ugly because what json:api requires us to do _is_ ugly. I could hide it, but then it would be less obvious that this thing is something I'm flagging to the developer as interesting/different/special.